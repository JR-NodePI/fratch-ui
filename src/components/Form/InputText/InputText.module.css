.input_text {
  position: relative;
  display: flex;
  box-shadow: var(--ft-box-shadow-1);
  border-radius: var(--ft-form-item-border-radius);
  background-color: var(--ft-color-grey-4);
  border: var(--ft-form-item-border);
}
.input_text .input {
  font-family: var(--ft-select-default-font-family);
  font-size: var(--ft-select-default-font-size);
  line-height: var(--ft-select-default-line-height);
  padding: var(--ft-form-item-padding-y) var(--ft-form-item-padding-x);
  width: 100%;
  display: block;
  box-sizing: border-box;
  color: var(--ft-color-text);
  border: none;
  border-radius: var(--ft-form-item-border-radius);
  background-color: transparent;
}
.input_text .input.cleanable {
  padding-right: calc(var(--ft-form-item-padding-x) * 3);
}
.input_text .input:focus {
  outline: var(--ft-select-list-item-bg-color-active) solid 2px;
}
.input_text .input:read-only {
  cursor: default;
  outline: none;
}
.input_text .input:disabled {
  background-color: var(--ft-color-grey-4);
  pointer-events: none;
  filter: saturate(0) brightness(115%) contrast(90%);
}

.input_text input::placeholder {
  color: var(--ft-color-grey-1);
}

/*** cleaner ***/
.input_text .cleaner {
  position: absolute;
  right: 0;
  border: none;
  margin: 0;
  padding: var(--ft-form-item-padding-y) 0;
  border-left: none;
  transition: background-color 0.2s ease-out;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: transparent;
  height: 100%;
  line-height: 100%;
  right: 0;
  width: calc(var(--ft-form-item-padding-x) * 3);
}
.input_text .cleaner > i {
  display: none;
}
.input_text .cleaner .cleaner_icon {
  color: var(--ft-form-item-border-color);
}
.input_text .cleaner:hover .cleaner_icon {
  color: var(--ft-color-text);
}
