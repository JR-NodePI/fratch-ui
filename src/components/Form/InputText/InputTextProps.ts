export interface InputTextProps {
  className?: string;
  cleanable?: boolean;
  cleanerClassName?: string;
  disabled?: boolean;
  id?: string;
  onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void;
  onChange?: (event?: React.ChangeEvent<HTMLInputElement>) => void;
  onClean?: (element?: HTMLInputElement) => void;
  onClick?: (event: React.MouseEvent<HTMLInputElement>) => void;
  onFocus?: (event: React.FocusEvent<HTMLInputElement>) => void;
  onKeyDownCapture?: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  placeholder?: string;
  readOnly?: boolean;
  title?: string;
  type?: 'text' | 'password' | 'email' | 'number' | 'tel' | 'url' | 'search';
  value?: string;
}
