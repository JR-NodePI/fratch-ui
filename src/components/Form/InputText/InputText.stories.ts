import type { Meta, StoryObj } from '@storybook/react';

import InputText from './InputText';

const meta = {
  title: 'Example/Form/InputText',
  component: InputText,
  tags: ['autodocs'],
  argTypes: { onChange: { action: 'onChange' } },
} satisfies Meta<typeof InputText>;

export default meta;
type Story = StoryObj<typeof InputText>;

export const Docs: Story = {
  args: {
    type: 'text',
    value: 'Input value',
  },
};
