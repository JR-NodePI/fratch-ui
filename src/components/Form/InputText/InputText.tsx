import { forwardRef, useEffect, useRef } from 'react';

import { c } from '../../../helpers/classNameHelpers';
import { IconClose } from '../../Icons/Icons';
import { type InputTextProps } from './InputTextProps';

import styles from './InputText.module.css';

const InputText = forwardRef<HTMLInputElement, InputTextProps>(
  (
    {
      className,
      cleanerClassName,
      disabled,
      onBlur,
      onChange,
      onClean,
      onClick,
      onFocus,
      onKeyDownCapture,
      placeholder,
      readOnly,
      type = 'text',
      value,
      title,
      cleanable,
    }: InputTextProps,
    ref
  ) => {
    const innerRef = useRef<HTMLInputElement>();

    useEffect(() => {
      if (innerRef.current != null) {
        innerRef.current.value = value ?? '';
      }
    }, [value]);

    const handleOnClean = (): void => {
      if (innerRef.current != null) {
        innerRef.current.value = '';
        innerRef.current.focus();
      }
      onChange?.();
      onClean?.(innerRef.current);
    };

    const setInnerRef = (current: HTMLInputElement) => {
      if (typeof ref === 'function') ref(current);
      else if (ref) ref.current = current;
      innerRef.current = current;
    };

    return (
      <div className={c(styles.input_wrapper, className)}>
        <input
          className={c(styles.input)}
          disabled={disabled}
          onBlur={onBlur}
          onChange={onChange}
          onClick={onClick}
          onFocus={onFocus}
          onKeyDownCapture={onKeyDownCapture}
          placeholder={placeholder}
          readOnly={readOnly}
          ref={setInnerRef}
          type={type}
          title={title}
        />
        {cleanable && (
          <button
            className={c(styles.cleaner, cleanerClassName)}
            onClick={handleOnClean}
          >
            <IconClose className={c(styles.cleaner_icon)} />
          </button>
        )}
      </div>
    );
  }
);

export default InputText;
