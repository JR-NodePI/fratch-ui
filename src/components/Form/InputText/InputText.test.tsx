import { createRef } from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import InputText from './InputText';
import { type InputTextProps } from './InputTextProps';

describe('InputText', () => {
  const initialProps: InputTextProps = {};

  const setup = (
    props = initialProps,
    ref?: React.ForwardedRef<HTMLInputElement>
  ) => render(<InputText ref={ref} {...props} />);

  it('should render properly', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('should render with id', () => {
    const { container } = setup({
      ...initialProps,
      id: 'mock-id',
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with title', () => {
    const { container } = setup({
      ...initialProps,
      title: 'mock-title',
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with type', () => {
    const { container } = setup({
      ...initialProps,
      type: 'password',
    });
    expect(container).toMatchSnapshot();
  });

  it('should be disabled', () => {
    setup({
      ...initialProps,
      disabled: true,
    });
    expect(screen.getByRole('textbox')).toBeDisabled();
  });

  it('should be readOnly', () => {
    setup({
      ...initialProps,
      readOnly: true,
    });
    expect(screen.getByRole('textbox')).toHaveAttribute('readonly');
  });

  it('should be cleanable, perform onClean and onChange', async () => {
    const onClean = vi.fn();
    const onChange = vi.fn();
    const mockInitialValue = 'mock-initial-value';

    const { container } = setup({
      ...initialProps,
      cleanable: true,
      value: mockInitialValue,
      cleanerClassName: 'mock-cleaner-class',
      onClean,
      onChange,
    });

    const textBoxElement = screen.getByRole('textbox');
    expect(textBoxElement).toHaveValue(mockInitialValue);

    await userEvent.click(screen.getByRole('button'));

    expect(textBoxElement).toHaveValue('');
    expect(textBoxElement).toHaveFocus();

    expect(onClean).toHaveBeenCalledTimes(1);
    expect(onClean).toHaveBeenCalledWith(textBoxElement);

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith();

    expect(container).toMatchSnapshot();
  });

  it('should perform onChange', async () => {
    const onChange = vi.fn();
    const mockInitialValue = 'mock-initial-value';

    setup({
      ...initialProps,
      value: mockInitialValue,
      onChange,
    });

    const textBoxElement = screen.getByRole('textbox');
    expect(textBoxElement).toHaveValue(mockInitialValue);

    await userEvent.type(textBoxElement, '1234');

    const mockEvent = expect.objectContaining({
      target: textBoxElement,
      type: 'change',
    });
    expect(onChange).toHaveBeenCalledTimes(4);
    expect(onChange).toHaveBeenCalledWith(mockEvent);
  });

  it('should perform onFocus and then perform onBlur', async () => {
    const onFocus = vi.fn();
    const onBlur = vi.fn();

    setup({
      ...initialProps,
      onFocus,
      onBlur,
    });

    const textBoxElement = screen.getByRole('textbox');
    await userEvent.click(textBoxElement);

    const mockEventFocus = expect.objectContaining({
      target: textBoxElement,
      type: 'focus',
    });
    expect(onFocus).toHaveBeenCalledTimes(1);
    expect(onFocus).toHaveBeenCalledWith(mockEventFocus);
    expect(textBoxElement).toHaveFocus();

    await userEvent.tab();

    const mockEventBlur = expect.objectContaining({
      target: textBoxElement,
      type: 'blur',
    });
    expect(onBlur).toHaveBeenCalledTimes(1);
    expect(onBlur).toHaveBeenCalledWith(mockEventBlur);
    expect(textBoxElement).not.toHaveFocus();
  });

  it('should perform onClick', async () => {
    const onClick = vi.fn();

    setup({
      ...initialProps,
      onClick,
    });

    const textBoxElement = screen.getByRole('textbox');
    await userEvent.click(textBoxElement);

    const mockEvent = expect.objectContaining({
      target: textBoxElement,
      type: 'click',
    });
    expect(onClick).toHaveBeenCalledTimes(1);
    expect(onClick).toHaveBeenCalledWith(mockEvent);
  });

  it('should perform onKeyDownCapture', async () => {
    const onKeyDownCapture = vi.fn();

    setup({
      ...initialProps,
      onKeyDownCapture,
    });

    const keyboardKeys = {
      A: 'A',
      a: 'a',
      '{enter}': 'Enter',
      '{backspace}': 'Backspace',
      '{delete}': 'Delete',
      '{tab}': 'Tab',
    };

    const keyboardKeyNames = Object.keys(keyboardKeys);
    const keyboardKeyValues = Object.values(keyboardKeys);

    const textBoxElement = screen.getByRole('textbox');
    await userEvent.click(textBoxElement);
    await userEvent.keyboard(keyboardKeyNames.join(''));

    expect(onKeyDownCapture).toHaveBeenCalledTimes(keyboardKeyValues.length);
    keyboardKeyValues.forEach(key => {
      const mockEvent = expect.objectContaining({
        target: textBoxElement,
        type: 'keydown',
        key,
      });
      expect(onKeyDownCapture).toHaveBeenCalledWith(mockEvent);
    });
  });

  it('should render with className', () => {
    const { container } = setup({
      ...initialProps,
      className: 'mock-class',
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with placeholder', () => {
    const { container } = setup({
      ...initialProps,
      placeholder: 'mock-placeholder',
    });
    expect(container).toMatchSnapshot();
  });

  it('should set the forwarded ref', async () => {
    const ref = createRef<HTMLInputElement>();
    setup(initialProps, ref);

    expect(ref).toMatchSnapshot();
  });
});
