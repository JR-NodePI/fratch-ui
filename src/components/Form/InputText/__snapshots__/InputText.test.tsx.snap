// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputText > should be cleanable, perform onClean and onChange 1`] = `
<div>
  <div
    class="_input_text_300ab4"
  >
    <input
      class="_input_300ab4"
      type="text"
    />
    <button
      class="_cleaner_300ab4 mock-cleaner-class"
    >
      <svg
        aria-hidden="true"
        class="_icon_2be8fc _primary_2be8fc _cleaner_icon_300ab4"
        fill="none"
        viewBox="-1 -1 16 16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`InputText > should render properly 1`] = `
<div>
  <div
    class="_input_text_300ab4"
  >
    <input
      class="_input_300ab4"
      type="text"
    />
  </div>
</div>
`;

exports[`InputText > should render with className 1`] = `
<div>
  <div
    class="_input_text_300ab4 mock-class"
  >
    <input
      class="_input_300ab4"
      type="text"
    />
  </div>
</div>
`;

exports[`InputText > should render with id 1`] = `
<div>
  <div
    class="_input_text_300ab4"
  >
    <input
      class="_input_300ab4"
      id="mock-id"
      name="mock-id"
      type="text"
    />
  </div>
</div>
`;

exports[`InputText > should render with placeholder 1`] = `
<div>
  <div
    class="_input_text_300ab4"
  >
    <input
      class="_input_300ab4"
      placeholder="mock-placeholder"
      type="text"
    />
  </div>
</div>
`;

exports[`InputText > should render with title 1`] = `
<div>
  <div
    class="_input_text_300ab4"
  >
    <input
      class="_input_300ab4"
      title="mock-title"
      type="text"
    />
  </div>
</div>
`;

exports[`InputText > should render with type 1`] = `
<div>
  <div
    class="_input_text_300ab4"
  >
    <input
      class="_input_300ab4"
      type="password"
    />
  </div>
</div>
`;

exports[`InputText > should set the forwarded ref 1`] = `
{
  "current": <input
    class="_input_300ab4"
    type="text"
  />,
}
`;
